sdl_viewer/src/graphic/tiled_texture_loader.rs

```diff
@@ -0,0 +1,107 @@
+use byteorder::{LittleEndian, ReadBytesExt};
+use image::{GenericImage, GenericImageView, ImageBuffer, LumaA, Pixel, Rgba};
+use num_integer::Integer;
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::BufReader;
+use std::path::Path;
+
+/// Little trait that just dispatches to the correct read_xyz_into call for the image type
+pub trait ReadLittleEndian {
+    fn read_from<R: ReadBytesExt>(&mut self, rdr: &mut R) -> Result<(), std::io::Error>;
+}
+
+impl ReadLittleEndian for ImageBuffer<LumaA<f32>, Vec<f32>> {
+    fn read_from<R: ReadBytesExt>(&mut self, rdr: &mut R) -> Result<(), std::io::Error> {
+        rdr.read_f32_into::<LittleEndian>(self)
+    }
+}
+
+impl ReadLittleEndian for ImageBuffer<Rgba<u8>, Vec<u8>> {
+    fn read_from<R: ReadBytesExt>(&mut self, rdr: &mut R) -> Result<(), std::io::Error> {
+        rdr.read_exact(self)
+    }
+}
+
+type TilePos = (i32, i32);
+
+pub struct TiledTextureLoader<P: Pixel> {
+    tile_size: u32,
+    tiles: HashMap<TilePos, ImageBuffer<P, Vec<P::Subpixel>>>,
+}
+
+impl<P> TiledTextureLoader<P>
+where
+    P: Pixel + 'static,
+    ImageBuffer<P, Vec<P::Subpixel>>: ReadLittleEndian,
+{
+    pub fn new<I, Q>(tile_size: u32, tile_iter: I) -> Result<Self, std::io::Error>
+    where
+        I: Iterator<Item = ((i32, i32), Q)>,
+        Q: AsRef<Path>,
+    {
+        let mut tiles = HashMap::new();
+        for (xy, file_path) in tile_iter {
+            let mut tile = ImageBuffer::new(tile_size, tile_size);
+            let mut tile_reader = BufReader::new(File::open(file_path)?);
+            tile.read_from(&mut tile_reader)?;
+            let previous_entry = tiles.insert(xy, tile);
+            assert!(previous_entry.is_none());
+        }
+
+        Ok(TiledTextureLoader { tile_size, tiles })
+    }
+
+    /// Loads the specified region of the sparse texture into a ImageBuffer
+    pub fn load(
+        &self,
+        min_x: i64,
```

Comment:
Wouldn't it make sense for `min_x` and `min_y` to have the same type as the indices of `self.tiles`? I.e. either all of them being `i32` or all of them `i64`. Also, do you allow negative numbers or could we use unsigned integers here?

Reply:
Negative positions are allowed, but we could make all of them `i64` for sure.